pkg/migrations/duplicate.go

```diff
@@ -17,174 +17,283 @@ import (
 // Duplicator duplicates a column in a table, including all constraints and
 // comments.
 type Duplicator struct {
+	stmtBuilder       *duplicatorStmtBuilder
 	conn              db.DB
-	table             *schema.Table
-	column            *schema.Column
-	asName            string
-	withoutNotNull    bool
-	withType          string
-	withoutConstraint string
+	columns           map[string]*columnToDuplicate
+	withoutConstraint []string
+}
+
+type columnToDuplicate struct {
+	column         *schema.Column
+	asName         string
+	withoutNotNull bool
+	withType       string
+}
+
+// duplicatorStmtBuilder is a helper for building SQL statements to duplicate
+// columns and constraints in a table.
+type duplicatorStmtBuilder struct {
+	table *schema.Table
 }
 
 const (
 	dataTypeMismatchErrorCode  pq.ErrorCode = "42804"
 	undefinedFunctionErrorCode pq.ErrorCode = "42883"
+
+	cCreateUniqueIndexSQL            = `CREATE UNIQUE INDEX CONCURRENTLY %s ON %s (%s)`
+	cSetDefaultSQL                   = `ALTER TABLE %s ALTER COLUMN %s SET DEFAULT %s`
+	cAlterTableAddCheckConstraintSQL = `ALTER TABLE %s ADD CONSTRAINT %s %s NOT VALID`
 )
 
 // NewColumnDuplicator creates a new Duplicator for a column.
 func NewColumnDuplicator(conn db.DB, table *schema.Table, column *schema.Column) *Duplicator {
 	return &Duplicator{
-		conn:     conn,
-		table:    table,
-		column:   column,
-		asName:   TemporaryName(column.Name),
-		withType: column.Type,
+		stmtBuilder: &duplicatorStmtBuilder{
+			table: table,
+		},
+		conn: conn,
+		columns: map[string]*columnToDuplicate{
+			column.Name: {
+				column:   column,
+				asName:   TemporaryName(column.Name),
+				withType: column.Type,
+			},
+		},
+		withoutConstraint: make([]string, 0),
+	}
+}
+
+func NewColumnGroupDuplicator(conn db.DB, table *schema.Table, columns []*schema.Column) *Duplicator {
```

Comment:
I think the `NewColumnDuplicator` and `NewColumnGroupDuplicator` constructor functions could be combined too.